api/fs/watcher.js

```diff
@@ -0,0 +1,249 @@
+import { EventEmitter } from '../events.js'
+import { FileHandle } from './handle.js'
+import { AbortError } from '../errors.js'
+import { rand64 } from '../crypto.js'
+import { Buffer } from '../buffer.js'
+import hooks from '../hooks.js'
+import ipc from '../ipc.js'
+import gc from '../gc.js'
+
+/**
+ * Encodes filename based on encoding preference.
+ * @ignore
+ * @param {Watcher} watcher
+ * @param {string} filename
+ * @return {string|Buffer}
+ */
+function encodeFilename (watcher, filename) {
+  if (!watcher.encoding || watcher.encoding === 'utf8') {
+    return filename.toString()
+  }
+
+  if (watcher.encoding === 'buffer') {
+    return Buffer.from(filename.toString())
+  }
+
+  return filename
+}
+
+/**
+ * Starts the `fs.Watcher`
+ * @ignore
+ * @param {Watcher} watcher
+ * @return {Promise}
+ */
+async function start (watcher) {
+  // throws if not accessible
+  await FileHandle.access(watcher.path)
+
+  if (!watcher.id || typeof watcher.id !== 'string') {
+    throw new TypeError('Expectig fs.Watcher to have a id.')
+  }
+
+  const result = await ipc.send('fs.watch', {
+    path: watcher.path,
+    id: watcher.id
+  })
+
+  if (result.err) {
+    throw result.err
+  }
+}
+
+/**
+ * Internal watcher data event listeer.
+ * @ignore
+ * @param {Watcher} watcher
+ * @return {function}
+ */
+function listen (watcher) {
+  return hooks.onData((event) => {
+    const { data, source } = event.detail.params
+    if (source !== 'fs.watch') {
+      return
+    }
+
+    if (BigInt(data.id) !== BigInt(watcher.id)) {
+      return
+    }
+
+    const { path, events } = data
+
+    watcher.emit('change', events[0], encodeFilename(watcher, path))
+  })
+}
+
+/**
+ * A container for a file system path watcher.
+ */
+export class Watcher extends EventEmitter {
+  /**
+   * The underlying `fs.Watcher` resource id.
+   * @ignore
+   * @type {string}
+   */
+  id = null
```

Comment:
Random TS question:

If you don't set this to null in the instance value field, and you guarantee set them in the constructor, typescript will let you type it as never null and just the value its typed as. Or maybe this is a strict null check option, I'm not sure. Typescript is a million variants of itself depending on how you configure it. 

What about not setting this to null initially, or if there is a reason, why do we set this to null here?

Reply:
just by habit
@chicoxyzzy what do you think we should set it to here?

Reply:
Not setting null here (unless I'm mistaken), and also setting in the constructor narrows the fields slightly to be a single type instead of also possibly null.